File: app/src/main/java/com/young/minor/livetool/utils/IActivityManagerProxy.java
package com.young.minor.livetool.utils;

import android.content.Intent;

import java.lang.reflect.InvocationHandler;
import java.lang.reflect.Method;

/**
 * created by yf on 2019/3/7.
 */
public class IActivityManagerProxy implements InvocationHandler {
    private Object mActivityManager;
    private static final String TAG = "IActivityManagerProxy";
    public IActivityManagerProxy(Object activityManager) {
        this.mActivityManager = activityManager;
    }
    @Override
    public Object invoke(Object o, Method method, Object[] args) throws Throwable {
        if ("startActivity".equals(method.getName())) {//1
            Intent intent = null;
            int index = 0;
            for (int i = 0; i < args.length; i++) {
                if (args[i] instanceof Intent) {
                    index = i;
                    break;
                }
            }
            intent = (Intent) args[index];
            Intent subIntent = new Intent();//2
            String packageName = "com.cgtn.minor.liveminority.ui.activity";
            subIntent.setClassName(packageName,packageName+".SubActivity");//3
            subIntent.putExtra(HookHelper.TARGET_INTENT, intent);//4
            args[index] = subIntent;//5
        }
        return method.invoke(mActivityManager, args);
    }
}

File: app/src/main/java/com/cgtn/minor/liveminority/utils/IActivityManagerProxy.kt
package com.cgtn.minor.liveminority.utils

import android.app.ActivityManager
import android.app.IntentService
import android.content.Intent
import java.lang.reflect.InvocationHandler
import java.lang.reflect.Method

class IActivityManagerProxy(activityManager: Any) :InvocationHandler{
    var mActivityManager:Any? = activityManager

    override fun invoke(proxy: Any?, method: Method?, args: Array<in Any>?): Any {
        if ("startActivity" == method!!.name){
            var intent:Intent?= null
            var index = 0
            for (i in 0..args!!.size){
                if (args[i] is Intent){
                    index = i
                    break
                }
            }
            intent = args[index] as Intent
            val subIntent = Intent()
            //todo 需要启动的Activity
            val packageName = "com.cgtn...."
            subIntent.setClassName(packageName, "$packageName.StubActivity")
            subIntent.putExtra(HookHelper.TARGET_INTENT,intent)
            args[index] = subIntent
        }
        return method.invoke(mActivityManager,args)
    }

}